block-storage/src/main/scala/coop/rchain/blockstorage/util/io/package.scala

```diff
@@ -4,36 +4,41 @@ import java.io.{EOFException, IOException}
 import java.nio.file._
 import java.util.stream.Collectors
 
-import cats.data.EitherT
 import cats.implicits._
 import cats.effect.Sync
-import coop.rchain.blockstorage.util.io.IOError.{IOErr, IOErrT}
+import coop.rchain.blockstorage.util.io.IOError.RaiseIOError
 
 import scala.collection.JavaConverters._
 
 package object io {
-  private[io] def handleIo[F[_]: Sync, A](
+  private[io] def handleIo[F[_]: Sync: RaiseIOError, A](
       io: => A,
       handleIoException: IOException => IOError
-  ): F[IOErr[A]] =
+  ): F[A] =
     Sync[F].delay {
       try {
-        Right(io)
+        io.pure[F]
       } catch {
-        case e: EOFException                  => Left(EndOfFile(e))
-        case e: IOException                   => Left(handleIoException(e))
-        case e: SecurityException             => Left(FileSecurityViolation(e))
-        case e: UnsupportedOperationException => Left(UnsupportedFileOperation(e))
-        case e: IllegalArgumentException      => Left(IllegalFileOperation(e))
-        case e                                => Left(UnexpectedIOError(e))
+        case e: EOFException =>
+          RaiseIOError[F].raise[A](EndOfFile(e))
+        case e: IOException =>
+          RaiseIOError[F].raise[A](handleIoException(e))
+        case e: SecurityException =>
+          RaiseIOError[F].raise[A](FileSecurityViolation(e))
+        case e: UnsupportedOperationException =>
+          RaiseIOError[F].raise[A](UnsupportedFileOperation(e))
+        case e: IllegalArgumentException =>
+          RaiseIOError[F].raise[A](IllegalFileOperation(e))
+        case e =>
+          RaiseIOError[F].raise[A](UnexpectedIOError(e))
```

Comment:
I think that this indirection actually makes the code less understandable. `RaiseIOError` is indeed `FunctorRaise[F, E]` but `E` is a subtype of `Exception` and concrete instance of this `FunctorRaise` is derived from the `Sync` itself. Is the only gain from using `FunctorRaise` (over `Sync`) is to use the least powerful abstraction?

cc @ArturGajowy

Reply:
`E` is not a subtype of `Exception` as `IOError` is not a subtype of `Exception`. Concrete instance of `RaiseIOError` is indeed being derived from `Sync` inside of `FileLMDBIndexBlockStore` constructor which wraps `IOError`s into a throwable class and uses `Sync[F].raiseError` on them. The main gain here is that if we decide to actually care about these IO errors (in other words, handle them in some meaningful way without crashing the node) we could just remove the automatic derivation from `Sync` and replace it with our own implementation of `FunctorRaise[F, E]`.

Reply:
Would this work with `monix.Task` though? Since it fixes any E to be subtype of `Throwable`?

Reply:
Would this work with `monix.Task` though? Since it fixes any E to be subtype of `Throwable`?

Reply:
> if we decide to actually care about these IO errors (in other words, handle them in some meaningful way without crashing the node)

you can still do that with current implementation (or any `Sync`).

Reply:
> Is the only gain from using `FunctorRaise` (over `Sync`) is to use the least powerful abstraction?

@goral09 are you suggesting this is not a sufficient reason? :smiling_imp:

> Would this work with `monix.Task` though?

Task has Sync, so yes.

> you can still do that with current implementation (or any `Sync`).

You're right. I think though @itegulov meant having some guidance from the typesystem.

Reply:
> > Would this work with `monix.Task` though?
> 
> Task has Sync, so yes.

That's not true. I asked whether using different error type (one that doesn't inherit from `Throwable`) would work. `Sync` if fixed for subtypes of `Throwable`.

> 
> > you can still do that with current implementation (or any `Sync`).
> 
> You're right. I think though @itegulov meant having some guidance from the typesystem.

I don't think this is true either. The only way to handle errors raised by the means of `FunctorRaise` is to recover from them using `handleError`, `handleErrorWith` etc. If you change the error to something that doesn't crash the node (as I understand sth that doesn't inherit from `THrowable`) you still have to handle errors the same way. So there's no gain.

Reply:
I thought you were asking if we can have `FunctorRaise[Task, IOError]`. Yes, we can. See the instance we provide based for any `F[_] : Sync`

Sure, with the errors being transferred through Sync.raise, we have no type guidance on where errors of which type can occur. It still makes sense to use `FunctorRaise` at the 'throw site', because we can then run with e.g. `F = EitehrT[IoError, Coeval]` or `F = zio.BIO` or whatever we want - and we need no changes to the 'throw site'.

As a general remark, please also mind that if a given solution doesn't meet your arbitrary ad-hoc made success criterion, it doesn't automatically mean there's no gain from it.

Reply:
> As a general remark, please also mind that if a given solution doesn't meet your arbitrary ad-hoc made success criterion, it doesn't automatically mean there's no gain from it.

I didn't say anywhere it doesn't meet my success criterion. I just expressed my opinion that in the case when there's a `Sync[F]` and `FunctorRaise[F, E]` available (and the fact the `F` will inevitably become a `monix.Task`) this gives less than it would have if `Sync` was bifunctor (or how should I call it) and `E` wasn't fixed for `_ <: Throwable`.

Reply:
As discussed in different comment, I think that using Defer + Applicative would partially solve this problem.